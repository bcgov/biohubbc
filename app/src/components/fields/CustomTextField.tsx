import TextField, { TextFieldProps } from '@material-ui/core/TextField';
import { useFormikContext } from 'formik';
import React from 'react';

export interface ICustomTextField {
  label: string;
  name: string;
  other?: TextFieldProps;
}

const CustomTextField: React.FC<ICustomTextField> = (props) => {
  const { touched, errors, values, handleChange } = useFormikContext<object>();
  const { name, label, other } = props;

  return (
    <TextField
      name={name}
      label={label}
      id={name}
      onChange={handleChange}
      variant="outlined"
      value={values[name]}
      fullWidth={true}
      error={touched[name] && Boolean(errors[name])}
      helperText={touched[name] && errors[name]}
      {...other}
    />
  );
};

export default CustomTextField;
