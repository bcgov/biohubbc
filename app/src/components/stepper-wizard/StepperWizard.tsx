import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
//import Container from '@material-ui/core/Container';
import Divider from '@material-ui/core/Divider';
import Paper from '@material-ui/core/Paper';
import Step from '@material-ui/core/Step';
import StepLabel from '@material-ui/core/StepLabel';
import Stepper from '@material-ui/core/Stepper';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import ArrowBack from '@material-ui/icons/ArrowBack';
import ArrowForward from '@material-ui/icons/ArrowForward';
import { mdiAlertCircle } from '@mdi/js';
import Icon from '@mdi/react';
import { Formik } from 'formik';
import React, { ReactElement, useCallback } from 'react';

const useStyles = makeStyles((theme) => ({
  actionButton: {
    minWidth: '6rem',
    '& + button': {
      marginLeft: '0.5rem'
    }
  },
  stepTitle: {
    marginBottom: 0
  },
  stepDescription: {
    maxWidth: '90ch'
  },
  stepper: {
    width: '280px',
    background: 'transparent'
  },
  [theme.breakpoints.down('md')]: {
    stepperContainer: {
      display: 'none'
    }
  }
}));

export interface IStepperWizardStep {
  stepTitle?: string;
  stepSubTitle?: string;
  stepContent: ReactElement;
  stepInitialValues: any;
  stepYupSchema?: object;
  isValid: boolean;
  isTouched: boolean;
}

export interface IStepperWizardProps {
  activeStep: number;
  steps: IStepperWizardStep[];
  innerRef: any;
  onNext: () => void;
  onPrevious: () => void;
  onSubmit: () => void;
  onSubmitLabel: string;
  onCancel: () => void;
  onChangeStep: (stepIndex: number) => void;
}

const StepperWizard: React.FC<IStepperWizardProps> = (props) => {
  const { activeStep, steps, innerRef, onChangeStep, onNext, onPrevious, onSubmit, onSubmitLabel, onCancel } = props;

  const classes = useStyles();

  const showPreviousButton = activeStep > 0;
  const showNextButton = activeStep < steps.length - 1;

  const getStepperWizardControls = () => {
    return (
      <>
        <Box>
          {showPreviousButton && (
            <Button
              startIcon={<ArrowBack color="primary" fontSize="small" />}
              variant="outlined"
              color="primary"
              onClick={onPrevious}
              className={classes.actionButton}
              data-testid="stepper_previous">
              Previous
            </Button>
          )}
        </Box>
        <Box>
          {showNextButton && (
            <Button
              endIcon={<ArrowForward fontSize="small" />}
              type="submit"
              variant="contained"
              color="primary"
              onClick={onNext}
              className={classes.actionButton}
              data-testid="stepper_next">
              Next
            </Button>
          )}
          <Button
            type="submit"
            variant="contained"
            color="primary"
            onClick={onSubmit}
            className={classes.actionButton}
            data-testid="stepper_submit">
            {onSubmitLabel}
          </Button>
          <Button
            variant="outlined"
            color="primary"
            onClick={onCancel}
            className={classes.actionButton}
            data-testid="stepper_cancel">
            Cancel
          </Button>
        </Box>
      </>
    );
  };

  const getStepperWizardLabels = () => {
    return steps.map((step, index) => {
      return (
        <Step key={index}>
          <StepLabel
            onClick={() => onChangeStep?.(index)}
            StepIconProps={{
              error: !step.isValid && step.isTouched,
              completed: step.isValid && step.isTouched,
              ...(!step.isValid &&
                step.isTouched && {
                  icon: <Icon path={mdiAlertCircle} className="MuiSvgIcon-root MuiStepIcon-root Mui-error" /> || {}
                })
            }}>
            <Typography variant="body2">{step.stepTitle}</Typography>
          </StepLabel>
        </Step>
      );
    });
  };

  const getStepperWizardContent = useCallback(() => {
    if (!steps?.length) {
      return;
    }

    return (
      <>
        <Box mb={3}>
          <Typography variant="h2" className={classes.stepTitle}>
            {steps[activeStep].stepTitle}
          </Typography>
        </Box>

        <Typography color="textSecondary" className={classes.stepDescription}>
          {steps[activeStep].stepSubTitle}
        </Typography>

        <Box>
          <Box mt={5}>
            <Formik
              key={steps[activeStep].stepTitle}
              innerRef={innerRef}
              enableReinitialize={true}
              initialValues={steps[activeStep].stepInitialValues}
              validationSchema={steps[activeStep].stepYupSchema}
              validateOnBlur={true}
              validateOnChange={false}
              onSubmit={() => {
                // do nothing
              }}>
              {steps[activeStep].stepContent}
            </Formik>
          </Box>
        </Box>
      </>
    );
  }, [activeStep, innerRef, steps, classes.stepTitle, classes.stepDescription]);

  return (
    <Box display="flex" component={Paper}>
      <Box className={classes.stepperContainer} display="flex">
        <Box p={5} pb={7}>
          <Stepper nonLinear orientation="vertical" activeStep={activeStep} className={classes.stepper}>
            {getStepperWizardLabels()}
          </Stepper>
        </Box>

        <Box>
          <Divider orientation="vertical" />
        </Box>
      </Box>

      <Box display="flex" flexDirection="column" flex="1 1 auto" p={5}>
        <Box flex="1 1 auto">{getStepperWizardContent()}</Box>
        <Box pt={7} pb={5}>
          <Divider orientation="horizontal" />
        </Box>
        <Box display="flex" justifyContent="space-between">
          {getStepperWizardControls()}
        </Box>
      </Box>
    </Box>
  );
};

export default StepperWizard;
