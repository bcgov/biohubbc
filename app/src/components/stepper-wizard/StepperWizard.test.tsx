import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import CustomTextField from 'components/fields/CustomTextField';
import StepperWizard, { IStepperWizardStep } from 'components/stepper-wizard/StepperWizard';
import React from 'react';
import yup from 'utils/YupSchema';

export interface ISampleFormikFormProps {
  testField: string;
}

export const SampleFormikFormYupSchema = yup.object().shape({
  testField: yup.string().required('You must provide a value for testField')
});

const SampleFormikForm = () => {
  return (
    <form>
      <CustomTextField name="testField" label="Test Field" other={{ multiline: true, required: true, rows: 4 }} />
    </form>
  );
};

const renderContainer = (
  activeStep: number,
  steps: IStepperWizardStep[],
  innerRef: any,
  onNext: () => void,
  onPrevious: () => void,
  onSubmit: () => void,
  onSubmitLabel: string,
  onCancel: () => void,
  onChangeStep: (stepIndex: number) => void
) => {
  return render(
    <StepperWizard
      activeStep={activeStep}
      steps={steps}
      innerRef={innerRef}
      onChangeStep={onChangeStep}
      onPrevious={onPrevious}
      onNext={onNext}
      onSubmit={onSubmit}
      onSubmitLabel={onSubmitLabel}
      onCancel={onCancel}
    />
  );
};

describe('StepperWizard', () => {
  describe('with 0 step forms', () => {
    it('renders correctly with no steps', () => {
      const innerRef = { current: null };
      const onNext = jest.fn();
      const onPrevious = jest.fn();
      const onSubmit = jest.fn();
      const onSubmitLabel = 'test submit label';
      const onCancel = jest.fn();
      const onChangeStep = jest.fn();

      const { asFragment } = renderContainer(
        0,
        [],
        innerRef,
        onNext,
        onPrevious,
        onSubmit,
        onSubmitLabel,
        onCancel,
        onChangeStep
      );

      // should render an essentially empty component
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with 2 step forms', () => {
    const innerRef = { current: null };
    const onNext = jest.fn();
    const onPrevious = jest.fn();
    const onSubmit = jest.fn();
    const onSubmitLabel = 'test submit label';
    const onCancel = jest.fn();
    const onChangeStep = jest.fn();

    afterEach(() => {
      jest.clearAllMocks();
      cleanup();
    });

    it('renders correctly with 2 steps', async () => {
      const { getAllByText, getByText } = renderContainer(
        0,
        [
          {
            stepTitle: 'Test Step Title 1',
            stepSubTitle: 'Test Step Description 1',
            stepContent: <SampleFormikForm />,
            stepInitialValues: {
              testfield: ''
            },
            stepYupSchema: SampleFormikFormYupSchema,
            isValid: false,
            isTouched: false
          },
          {
            stepTitle: 'Test Step Title 2',
            stepSubTitle: 'Test Step Description 2',
            stepContent: <SampleFormikForm />,
            stepInitialValues: {
              testfield: ''
            },
            stepYupSchema: SampleFormikFormYupSchema,
            isValid: false,
            isTouched: false
          }
        ],
        innerRef,
        onNext,
        onPrevious,
        onSubmit,
        onSubmitLabel,
        onCancel,
        onChangeStep
      );

      await waitFor(() => {
        expect(getAllByText('Test Step Title 1').length).toEqual(2);
        expect(getByText('Test Step Description 1')).toBeVisible();

        expect(getByText('Test Step Title 2')).toBeVisible();
      });
    });

    it('calls onSubmit prop on submit button click', async () => {
      const { getByTestId } = renderContainer(
        0,
        [
          {
            stepTitle: 'Test Step Title 1',
            stepSubTitle: 'Test Step Description 1',
            stepContent: <SampleFormikForm />,
            stepInitialValues: {
              testfield: ''
            },
            stepYupSchema: SampleFormikFormYupSchema,
            isValid: true,
            isTouched: false
          },
          {
            stepTitle: 'Test Step Title 2',
            stepSubTitle: 'Test Step Description 2',
            stepContent: <SampleFormikForm />,
            stepInitialValues: {
              testfield: ''
            },
            stepYupSchema: SampleFormikFormYupSchema,
            isValid: false,
            isTouched: false
          }
        ],
        innerRef,
        onNext,
        onPrevious,
        onSubmit,
        onSubmitLabel,
        onCancel,
        onChangeStep
      );

      fireEvent.click(getByTestId('stepper_submit'));

      await waitFor(() => {
        expect(onSubmit).toHaveBeenCalledTimes(1);
      });
    });

    it('calls onNext props on next button click', async () => {
      const { getByTestId } = renderContainer(
        0,
        [
          {
            stepTitle: 'Test Step Title 1',
            stepSubTitle: 'Test Step Description 1',
            stepContent: <SampleFormikForm />,
            stepInitialValues: {
              testfield: ''
            },
            stepYupSchema: SampleFormikFormYupSchema,
            isValid: false,
            isTouched: true
          },
          {
            stepTitle: 'Test Step Title 2',
            stepSubTitle: 'Test Step Description 2',
            stepContent: <SampleFormikForm />,
            stepInitialValues: {
              testfield: ''
            },
            stepYupSchema: SampleFormikFormYupSchema,
            isValid: false,
            isTouched: false
          }
        ],
        innerRef,
        onNext,
        onPrevious,
        onSubmit,
        onSubmitLabel,
        onCancel,
        onChangeStep
      );

      fireEvent.click(getByTestId('stepper_next'));

      await waitFor(() => {
        expect(onNext).toHaveBeenCalledTimes(1);
      });
    });

    it('calls onPrevious prop on previous button click', async () => {
      const { getByTestId } = renderContainer(
        1,
        [
          {
            stepTitle: 'Test Step Title 1',
            stepSubTitle: 'Test Step Description 1',
            stepContent: <SampleFormikForm />,
            stepInitialValues: {
              testfield: ''
            },
            stepYupSchema: SampleFormikFormYupSchema,
            isValid: true,
            isTouched: true
          },
          {
            stepTitle: 'Test Step Title 2',
            stepSubTitle: 'Test Step Description 2',
            stepContent: <SampleFormikForm />,
            stepInitialValues: {
              testfield: ''
            },
            stepYupSchema: SampleFormikFormYupSchema,
            isValid: false,
            isTouched: false
          }
        ],
        innerRef,
        onNext,
        onPrevious,
        onSubmit,
        onSubmitLabel,
        onCancel,
        onChangeStep
      );

      fireEvent.click(getByTestId('stepper_previous'));

      await waitFor(() => {
        expect(onPrevious).toHaveBeenCalledTimes(1);
      });
    });

    it('calls onCancel prop on cancel button click', async () => {
      const { getByTestId } = renderContainer(
        0,
        [
          {
            stepTitle: 'Test Step Title 1',
            stepSubTitle: 'Test Step Description 1',
            stepContent: <SampleFormikForm />,
            stepInitialValues: {
              testfield: ''
            },
            stepYupSchema: SampleFormikFormYupSchema,
            isValid: true,
            isTouched: true
          },
          {
            stepTitle: 'Test Step Title 2',
            stepSubTitle: 'Test Step Description 2',
            stepContent: <SampleFormikForm />,
            stepInitialValues: {
              testfield: ''
            },
            stepYupSchema: SampleFormikFormYupSchema,
            isValid: true,
            isTouched: true
          }
        ],
        innerRef,
        onNext,
        onPrevious,
        onSubmit,
        onSubmitLabel,
        onCancel,
        onChangeStep
      );

      fireEvent.click(getByTestId('stepper_cancel'));

      await waitFor(() => {
        expect(onCancel).toHaveBeenCalledTimes(1);
      });
    });
  });
});
