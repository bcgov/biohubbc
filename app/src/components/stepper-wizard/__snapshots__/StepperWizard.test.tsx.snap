// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepperWizard with 0 step forms renders correctly with no steps 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiBox-root MuiBox-root-6 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiBox-root MuiBox-root-7 makeStyles-stepperContainer-5"
    >
      <div
        class="MuiBox-root MuiBox-root-8"
      >
        <div
          class="MuiPaper-root MuiStepper-root MuiStepper-vertical makeStyles-stepper-4 MuiPaper-elevation0"
        />
      </div>
      <div
        class="MuiBox-root MuiBox-root-9"
      >
        <hr
          class="MuiDivider-root MuiDivider-vertical"
        />
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-10"
    >
      <div
        class="MuiBox-root MuiBox-root-11"
      />
      <div
        class="MuiBox-root MuiBox-root-12"
      >
        <hr
          class="MuiDivider-root"
        />
      </div>
      <div
        class="MuiBox-root MuiBox-root-13"
      >
        <div
          class="MuiBox-root MuiBox-root-14"
        />
        <div
          class="MuiBox-root MuiBox-root-15"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-actionButton-1 MuiButton-containedPrimary"
            data-testid="stepper_submit"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              test submit label
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-actionButton-1 MuiButton-outlinedPrimary"
            data-testid="stepper_cancel"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Cancel
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
